File: LinkedList/MiddleOfTheLinkedList.cpp
//  # Definition for singly-linked list.
struct ListNode {
  int val;
  ListNode *next;
  ListNode(int x) : val(x), next(NULL) {}
};

class Solution
{
public:
    ListNode* middleNode(ListNode* head)
    {
        ListNode* tmp = head;
        vector<int> buf;
        while(tmp)
        {
            buf.push_back(tmp->val);
            tmp = tmp->next;
        }
        int index = ceil((float)(buf.size()-1.0)/2);
        int i = 0;
        ListNode* tmp2 = head;
        while(i<index)
        {
            tmp2=tmp2->next;
            i++;
        }
        return tmp2;
    }
};

File: LinkedList/MiddleOfTheLinkedList.py
# -*- coding:utf-8 -*-
"""
author:Bonnieliuliu
email:fangyuanliu@pku.edu.cn

file: MiddleOfTheLinkedList.py
time: 2018/8/6 2:01
"""
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def middleNode(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        tmp = head
        buf = []
        while(tmp):
            buf.append(tmp.val)
            tmp = tmp.next
        if len(buf) == 1:
            return head
        index = int((len(buf)-2)/2)+1
        i = 0
        tmp2 = head
        while(i<index):
            tmp2 = tmp2.next
            i += 1
        return tmp2